Data/HashMap/Array.hs

```diff
@@ -232,6 +234,13 @@ run :: (forall s . ST s (MArray s e)) -> Array e
 run act = runST $ act >>= unsafeFreeze
 {-# INLINE run #-}
 
+runInternal :: (forall s . ST s (Int, MArray s e)) -> (Int, Array e)
```

Comment:
To help GHC unbox this, and to reduce the cost of GHC *failing* to unbox this, I suggest you consider

```haskell
data RunRes e = RunRes !Int !(Array e)
```

instead of `(Int, Array e)`. Note that GHC is generally not all that great at unboxing "deeply". Here you're asking it to unbox `(# State# s, (Int, Array e) #)`, where `Int` and `Array e` are themselves boxed. With my suggestion, that's cut down a bit, with the `Int` becoming an `Int#` and the `Array e` becoming an `Array# e`.

Reply:
I'll fix this the way you suggest.